File: gpio/python/gpio_controller.py
# Copyright (c) 2025 by T3 Foundation. All rights reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     https://www.apache.org/licenses/LICENSE-2.0
#     https://docs.t3gemstone.org/en/license
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#
# SPDX-License-Identifier: Apache-2.0

import sys
import time
from typing import Optional

import gpiod


class GpioController:
    def __init__(self):
        self.m_is_running: bool = False
        self.m_chip1: Optional[gpiod.Chip] = None

        self.m_output_request: Optional[gpiod.LineRequest] = None
        self.m_input_request: Optional[gpiod.LineRequest] = None

        self.m_line_gpio27: int = 33  # GPIO27 set to active-high output with low value
        self.m_line_led_red: int = 11  # LED_RED output GPIO
        self.m_line_led_green: int = 12  # LED_GREEN output GPIO
        self.m_line_gpio22: int = 41  # GPIO22 set to input with pull-up resistor enabled (normally high)

        self.m_prev_input_state: int = 0
        self.m_current_input_state: int = 0

        self._cleaned_up: bool = False

    def __del__(self):
        self.cleanup()

    def initialize(self) -> int:
        try:
            self.m_chip1 = gpiod.Chip("/dev/gpiochip1")

            output_config = {
                self.m_line_gpio27: gpiod.LineSettings(
                    direction=gpiod.line.Direction.OUTPUT, output_value=gpiod.line.Value.INACTIVE
                ),
                self.m_line_led_red: gpiod.LineSettings(
                    direction=gpiod.line.Direction.OUTPUT, active_low=True, output_value=gpiod.line.Value.INACTIVE
                ),
                self.m_line_led_green: gpiod.LineSettings(
                    direction=gpiod.line.Direction.OUTPUT, output_value=gpiod.line.Value.INACTIVE
                ),
            }

            self.m_output_request = self.m_chip1.request_lines(consumer="gpio_example", config=output_config)

            input_config = {
                self.m_line_gpio22: gpiod.LineSettings(
                    direction=gpiod.line.Direction.INPUT, bias=gpiod.line.Bias.PULL_UP
                )
            }

            self.m_input_request = self.m_chip1.request_lines(consumer="gpio_example", config=input_config)

        except Exception as e:
            print(f"Failed to initialize GPIO: {e}", file=sys.stderr)
            return 1

        # Read initial state of input
        try:
            self.m_prev_input_state = self.m_input_request.get_value(self.m_line_gpio22).value
        except Exception as e:
            print(f"Failed to read initial input state: {e}", file=sys.stderr)
            return 1

        self._print_configuration()
        return 0

    def _print_configuration(self) -> None:
        print("GPIO configuration complete:")
        print("- gpiochip1-33 (GPIO27)   : active-high output, value=0")
        print("- gpiochip1-11 (RED LED)  : active-low output , value=0")
        print("- gpiochip1-12 (GREEN LED): active-high output, value=0")
        print("- gpiochip1-41 (GPIO22)  : pull-up input")
        print()
        print("Waiting for input transitions on GPIO22...")
        print("Press Ctrl+C to exit")
        print()

    def run(self) -> None:
        try:
            self.m_is_running = True
            while self.m_is_running:
                try:
                    self.m_current_input_state = self.m_input_request.get_value(self.m_line_gpio22).value
                except Exception as e:
                    print(f"Failed to read input state: {e}", file=sys.stderr)
                    break

                if self.m_prev_input_state == 1 and self.m_current_input_state == 0:
                    values = {
                        self.m_line_led_red: gpiod.line.Value.ACTIVE,
                        self.m_line_led_green: gpiod.line.Value.INACTIVE,
                    }
                    self.m_output_request.set_values(values)
                    print("-> Set LED_RED=HIGH, LED_GREEN=LOW")
                elif self.m_prev_input_state == 0 and self.m_current_input_state == 1:
                    values = {
                        self.m_line_led_red: gpiod.line.Value.INACTIVE,
                        self.m_line_led_green: gpiod.line.Value.ACTIVE,
                    }
                    self.m_output_request.set_values(values)
                    print("-> Set LED_RED=LOW, LED_GREEN=HIGH")

                self.m_prev_input_state = self.m_current_input_state

                # Small delay to avoid excessive CPU usage
                time.sleep(0.01)

        except KeyboardInterrupt:
            print("\nReceived interrupt signal")
        except Exception as e:
            print(f"Error in main loop: {e}", file=sys.stderr)

    def stop(self) -> None:
        self.m_is_running = False

    def cleanup(self) -> None:
        if self._cleaned_up:
            return

        self._cleaned_up = True

        if self.m_output_request:
            self.m_output_request.release()
            self.m_output_request = None

        if self.m_input_request:
            self.m_input_request.release()
            self.m_input_request = None

File: gpio/cpp/gpio_controller.cpp
// Copyright (c) 2025 by T3 Foundation. All rights reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     https://www.apache.org/licenses/LICENSE-2.0
//     https://www.t3gemstone.org/license
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// SPDX-License-Identifier: Apache-2.0

#include "gpio_controller.h"
#include <iostream>
#include <threads.h>
#include <unistd.h>

GpioController::GpioController() {}

GpioController::~GpioController()
{
    cleanup();
}

void GpioController::delay_ms(int ms)
{
    struct timespec request = {ms / 1000, (ms % 1000) * 1'000'000};
    struct timespec remaining;

    while (thrd_sleep(&request, &remaining) == -1 && errno == EINTR)
    {
        request = remaining; // Sleep again with remaining time if interrupted
    }
}

bool GpioController::initialize()
{
    m_chip1 = gpiod_chip_open_by_name("gpiochip1");
    if (!m_chip1)
    {
        std::cerr << "Failed to open gpiochip1" << std::endl;
        return false;
    }

    m_chip2 = gpiod_chip_open_by_name("gpiochip2");
    if (!m_chip2)
    {
        std::cerr << "Failed to open gpiochip2" << std::endl;
        return false;
    }

    m_line_gpio4 = gpiod_chip_get_line(m_chip1, 38);
    m_line_led_red = gpiod_chip_get_line(m_chip1, 11);
    m_line_led_green = gpiod_chip_get_line(m_chip1, 12);
    m_line_gpio17 = gpiod_chip_get_line(m_chip2, 8);

    if (!m_line_gpio4 || !m_line_led_red || !m_line_led_green || !m_line_gpio17)
    {
        std::cerr << "Failed to get GPIO lines" << std::endl;
        return false;
    }

    if (!configure_outputs() || !configure_inputs())
    {
        return false;
    }

    // Read initial state of input
    m_prev_input_state = gpiod_line_get_value(m_line_gpio17);
    if (m_prev_input_state < 0)
    {
        std::cerr << "Failed to read initial input state" << std::endl;
        return false;
    }

    print_configuration();

    return true;
}

bool GpioController::configure_outputs()
{
    // Configure gpiochip1-38 as active-high output with value 0
    int ret = gpiod_line_request_output(m_line_gpio4, "gpio_example", 0);
    if (ret < 0)
    {
        std::cerr << "Failed to configure line1-38 as output" << std::endl;
        return false;
    }

    // Configure gpiochip1-11 as active-low output with value 0
    ret = gpiod_line_request_output_flags(m_line_led_red, "gpio_example", GPIOD_LINE_REQUEST_FLAG_ACTIVE_LOW, 0);
    if (ret < 0)
    {
        std::cerr << "Failed to configure line1-11 as active-low output" << std::endl;
        return false;
    }

    // Configure gpiochip1-12 as active-high output with value 0
    ret = gpiod_line_request_output(m_line_led_green, "gpio_example", 0);
    if (ret < 0)
    {
        std::cerr << "Failed to configure line1-12 as output" << std::endl;
        return false;
    }

    return true;
}

bool GpioController::configure_inputs()
{
    // Configure gpiochip2-8 as pull-up input
    int ret = gpiod_line_request_input_flags(m_line_gpio17, "gpio_example", GPIOD_LINE_REQUEST_FLAG_BIAS_PULL_UP);
    if (ret < 0)
    {
        std::cerr << "Failed to configure line2-8 as input" << std::endl;
        return false;
    }

    return true;
}

void GpioController::print_configuration()
{
    std::cout << "GPIO configuration complete:" << std::endl;
    std::cout << "- gpiochip1-38 (GPIO4)    : active-high output, value=0" << std::endl;
    std::cout << "- gpiochip1-11 (RED LED)  : active-low output , value=0" << std::endl;
    std::cout << "- gpiochip1-12 (GREEN LED): active-high output, value=0" << std::endl;
    std::cout << "- gpiochip2-8  (GPIO17)   : pull-up input" << std::endl;
    std::cout << std::endl;
    std::cout << "Waiting for input transitions on GPIO17..." << std::endl;
    std::cout << "Press Ctrl+C to exit" << std::endl << std::endl;
}

void GpioController::run()
{
    while (true)
    {
        m_current_input_state = gpiod_line_get_value(m_line_gpio17);
        if (m_current_input_state < 0)
        {
            std::cerr << "Failed to read input state" << std::endl;
            break;
        }

        if (m_prev_input_state == 1 && m_current_input_state == 0)
        {
            int ret = gpiod_line_set_value(m_line_led_red, 1);
            if (ret < 0)
            {
                std::cerr << "Failed to set LED_RED" << std::endl;
                break;
            }
            ret = gpiod_line_set_value(m_line_led_green, 0);
            if (ret < 0)
            {
                std::cerr << "Failed to set LED_GREEN" << std::endl;
                break;
            }
            std::cout << "-> Set LED_RED=HIGH, LED_GREEN=LOW" << std::endl;
        }
        if (m_prev_input_state == 0 && m_current_input_state == 1)
        {
            int ret = gpiod_line_set_value(m_line_led_red, 0);
            if (ret < 0)
            {
                std::cerr << "Failed to set LED_RED" << std::endl;
                break;
            }
            ret = gpiod_line_set_value(m_line_led_green, 1);
            if (ret < 0)
            {
                std::cerr << "Failed to set LED_GREEN" << std::endl;
                break;
            }
            std::cout << "-> Set LED_RED=LOW, LED_GREEN=HIGH" << std::endl;
        }

        m_prev_input_state = m_current_input_state;

        // Small delay to avoid excessive CPU usage
        delay_ms(10);
    }
}

void GpioController::cleanup()
{
    std::cout << "\nCleaning up..." << std::endl;

    if (m_line_gpio4)
    {
        gpiod_line_release(m_line_gpio4);
        m_line_gpio4 = nullptr;
    }
    if (m_line_led_red)
    {
        gpiod_line_release(m_line_led_red);
        m_line_led_red = nullptr;
    }
    if (m_line_led_green)
    {
        gpiod_line_release(m_line_led_green);
        m_line_led_green = nullptr;
    }
    if (m_line_gpio17)
    {
        gpiod_line_release(m_line_gpio17);
        m_line_gpio17 = nullptr;
    }

    if (m_chip1)
    {
        gpiod_chip_close(m_chip1);
        m_chip1 = nullptr;
    }
    if (m_chip2)
    {
        gpiod_chip_close(m_chip2);
        m_chip2 = nullptr;
    }
}
